File: src/main/java/cn/xihan/lib/easyhttp/callback/DownloadCallback.java
package cn.xihan.lib.easyhttp.callback;


import cn.xihan.lib.easyhttp.EasyLog;
import cn.xihan.lib.easyhttp.EasyUtils;
import cn.xihan.lib.easyhttp.config.IRequestInterceptor;
import cn.xihan.lib.easyhttp.exception.FileMD5Exception;
import cn.xihan.lib.easyhttp.exception.NullBodyException;
import cn.xihan.lib.easyhttp.exception.ResponseException;

import cn.xihan.lib.easyhttp.listener.OnDownloadListener;
import cn.xihan.lib.easyhttp.request.HttpRequest;
import okhttp3.Call;
import okhttp3.Response;
import okhttp3.ResponseBody;

import java.io.File;
import java.io.InputStream;
import java.io.OutputStream;

/**
 *    author : Android 轮子哥
 *    github : <a href="https://github.com/getActivity/EasyHttp">...</a>
 *    time   : 2019/11/25
 *    desc   : 下载接口回调
 */
public final class DownloadCallback extends BaseCallback {

    /** 请求配置 */
    private final HttpRequest<?> mHttpRequest;

    /** 文件 MD5 正则表达式 */
    private static final String FILE_MD5_REGEX = "^[\\w]{32}$";

    /** 保存的文件 */
    private File mFile;

    /** 校验的 MD5 */
    private String mMd5;

    /** 下载监听回调 */
    private OnDownloadListener mListener;

    /** 下载总字节 */
    private long mTotalByte;

    /** 已下载字节 */
    private long mDownloadByte;

    /** 下载进度 */
    private int mDownloadProgress;

    public DownloadCallback(HttpRequest<?> request) {
        super(request);
        mHttpRequest = request;
    }

    public DownloadCallback setFile(File file) {
        mFile = file;
        return this;
    }

    public DownloadCallback setMd5(String md5) {
        mMd5 = md5;
        return this;
    }

    public DownloadCallback setListener(OnDownloadListener listener) {
        mListener = listener;
        return this;
    }

    @Override
    protected void onStart(Call call) {
        EasyUtils.runOnAssignThread(mHttpRequest.getThreadSchedulers(), this::callOnStart);
    }

    @Override
    protected void onResponse(Response response) throws Exception {
        // 打印请求耗时时间
        EasyLog.printLog(mHttpRequest, "RequestConsuming：" +
                (response.receivedResponseAtMillis() - response.sentRequestAtMillis()) + " ms");

        IRequestInterceptor interceptor = mHttpRequest.getRequestInterceptor();
        if (interceptor != null) {
            response = interceptor.interceptResponse(mHttpRequest, response);
        }

        if (!response.isSuccessful())  {
            throw new ResponseException("The request failed, responseCode: " +
                    response.code() + ", message: " + response.message(), response);
        }

        // 如果没有指定文件的 md5 值
        if (mMd5 == null) {
            // 获取响应头中的文件 MD5 值
            String md5 = response.header("Content-MD5");
            // 这个 md5 值必须是文件的 md5 值
            if (md5 != null && md5.matches(FILE_MD5_REGEX)) {
                mMd5 = md5;
            }
        }

        File parentFile = mFile.getParentFile();
        if (parentFile != null) {
            EasyUtils.createFolder(parentFile);
        }
        ResponseBody body = response.body();
        if (body == null) {
            throw new NullBodyException("The response body is empty");
        }

        mTotalByte = body.contentLength();
        if (mTotalByte < 0) {
            mTotalByte = 0;
        }

        // 如果这个文件已经下载过，并且经过校验 MD5 是同一个文件的话，就直接回调下载成功监听
        if (!EasyUtils.isEmpty(mMd5) && mFile.isFile() &&
                mMd5.equalsIgnoreCase(EasyUtils.getFileMd5(EasyUtils.openFileInputStream(mFile)))) {
            // 文件已存在，跳过下载
            EasyLog.printLog(mHttpRequest, mFile.getPath() + " file already exists, skip download");
            EasyUtils.runOnAssignThread(mHttpRequest.getThreadSchedulers(), () -> callOnComplete(true));
            return;
        }

        int readLength;
        mDownloadByte = 0;
        byte[] bytes = new byte[8192];
        InputStream inputStream = body.byteStream();
        OutputStream outputStream = EasyUtils.openFileOutputStream(mFile);
        while ((readLength = inputStream.read(bytes)) != -1) {
            mDownloadByte += readLength;
            outputStream.write(bytes, 0, readLength);
            EasyUtils.runOnAssignThread(mHttpRequest.getThreadSchedulers(), this::callOnProgress);
        }
        EasyUtils.closeStream(inputStream);
        EasyUtils.closeStream(outputStream);

        String md5 = EasyUtils.getFileMd5(EasyUtils.openFileInputStream(mFile));
        if (!EasyUtils.isEmpty(mMd5) && !mMd5.equalsIgnoreCase(md5)) {
            // 文件 MD5 值校验失败
            throw new FileMD5Exception("MD5 verify failure", md5);
        }

        EasyLog.printLog(mHttpRequest, mFile.getPath() + " download completed");
        EasyUtils.runOnAssignThread(mHttpRequest.getThreadSchedulers(), () -> callOnComplete(false));
    }

    @Override
    protected void onFailure(final Exception e) {
        EasyLog.printThrowable(mHttpRequest, e);
        // 打印错误堆栈
        final Exception finalException = mHttpRequest.getRequestHandler().downloadFail(mHttpRequest, e);
        if (finalException != e) {
            EasyLog.printThrowable(mHttpRequest, finalException);
        }

        EasyLog.printLog(mHttpRequest, mFile.getPath() + " download error");
        EasyUtils.runOnAssignThread(mHttpRequest.getThreadSchedulers(), () -> callOnError(finalException));
    }

    private void callOnStart() {
        if (mListener == null) {
            return;
        }
        mListener.onStart(mFile);
    }

    private void callOnProgress() {
        if (mListener == null) {
            return;
        }
        mListener.onByte(mFile, mTotalByte, mDownloadByte);
        int progress = EasyUtils.getProgressProgress(mTotalByte, mDownloadByte);
        // 只有下载进度发生改变的时候才回调此方法，避免引起不必要的 View 重绘
        if (progress != mDownloadProgress) {
            mDownloadProgress = progress;
            mListener.onProgress(mFile, mDownloadProgress);
            EasyLog.printLog(mHttpRequest, mFile.getPath() +
                    ", downloaded: " + mDownloadByte + " / " + mTotalByte +
                    ", progress: " + progress + " %");
        }
    }

    private void callOnComplete(boolean cache) {
        if (mListener == null) {
            return;
        }
        mListener.onComplete(mFile, cache);
        mListener.onEnd(mFile);
    }

    private void callOnError(Exception e) {
        if (mListener == null) {
            return;
        }
        mListener.onError(mFile, e);
        mListener.onEnd(mFile);
    }
}

File: src/main/java/cn/xihan/lib/easyhttp/callback/DownloadCallback.kt
package cn.xihan.lib.easyhttp.callback


import cn.xihan.lib.easyhttp.EasyLog.printLog
import cn.xihan.lib.easyhttp.EasyLog.printThrowable
import cn.xihan.lib.easyhttp.EasyUtils.closeStream
import cn.xihan.lib.easyhttp.EasyUtils.createFolder
import cn.xihan.lib.easyhttp.EasyUtils.getFileMd5
import cn.xihan.lib.easyhttp.EasyUtils.getProgressProgress
import cn.xihan.lib.easyhttp.EasyUtils.isEmpty
import cn.xihan.lib.easyhttp.EasyUtils.openFileInputStream
import cn.xihan.lib.easyhttp.EasyUtils.openFileOutputStream
import cn.xihan.lib.easyhttp.EasyUtils.post
import cn.xihan.lib.easyhttp.exception.MD5Exception
import cn.xihan.lib.easyhttp.exception.NullBodyException
import cn.xihan.lib.easyhttp.listener.OnDownloadListener
import cn.xihan.lib.easyhttp.request.HttpRequest
import okhttp3.Call
import okhttp3.Response
import java.io.File

/**
 * author : Android 轮子哥
 * github : [https://github.com/getActivity/EasyHttp](https://github.com/getActivity/EasyHttp)
 * time   : 2019/11/25
 * desc   : 下载接口回调
 */
class DownloadCallback(
    /** 请求配置  */
    private val mHttpRequest: HttpRequest<*>
) : BaseCallback(mHttpRequest) {
    /** 保存的文件  */
    private var mFile: File? = null

    /** 校验的 MD5  */
    private var mMd5: String? = null

    /** 下载监听回调  */
    private var mListener: OnDownloadListener? = null

    /** 下载总字节  */
    private var mTotalByte: Long = 0

    /** 已下载字节  */
    private var mDownloadByte: Long = 0

    /** 下载进度  */
    private var mDownloadProgress = 0
    fun setFile(file: File?): DownloadCallback {
        mFile = file
        return this
    }

    fun setMd5(md5: String?): DownloadCallback {
        mMd5 = md5
        return this
    }

    fun setListener(listener: OnDownloadListener?): DownloadCallback {
        mListener = listener
        return this
    }

    override fun onStart(call: Call?) {
        post {
            if (mListener == null) {
                return@post
            }
            mListener!!.onStart(mFile)
        }
    }

    @Throws(Exception::class)
    override fun onResponse(response: Response?) {
        // 打印请求耗时时间
        var response = response
        printLog(
            mHttpRequest, "RequestConsuming：" +
                    (response!!.receivedResponseAtMillis - response.sentRequestAtMillis) + " ms"
        )
        val interceptor = mHttpRequest.requestInterceptor
        if (interceptor != null) {
            response = interceptor.interceptResponse(mHttpRequest, response)
        }

        // 如果没有指定文件的 md5 值
        if (mMd5 == null) {
            // 获取响应头中的文件 MD5 值
            val md5 = response!!.header("Content-MD5")
            // 这个 md5 值必须是文件的 md5 值
            if (!isEmpty(md5) && md5!!.matches(Regex(FILE_MD5_REGEX))) {
                mMd5 = md5
            }
        }
        val parentFile = mFile!!.parentFile
        if (parentFile != null) {
            createFolder(parentFile)
        }
        val body = response!!.body ?: throw NullBodyException("The response body is empty")
        mTotalByte = body.contentLength()
        if (mTotalByte < 0) {
            mTotalByte = 0
        }

        // 如果这个文件已经下载过，并且经过校验 MD5 是同一个文件的话，就直接回调下载成功监听
        if (!isEmpty(mMd5) && mFile!!.isFile &&
            mMd5.equals(getFileMd5(openFileInputStream(mFile!!)), ignoreCase = true)
        ) {
            post {
                if (mListener == null) {
                    return@post
                }
                mListener!!.onComplete(mFile, true)
                mListener!!.onEnd(mFile)
                // 文件已存在，跳过下载
                printLog(mHttpRequest, mFile!!.path + " file already exists, skip download")
            }
            return
        }
        var readLength: Int
        mDownloadByte = 0
        val bytes = ByteArray(8192)
        val inputStream = body.byteStream()
        val outputStream = openFileOutputStream(mFile!!)
        while (inputStream.read(bytes).also { readLength = it } != -1) {
            mDownloadByte += readLength.toLong()
            outputStream.write(bytes, 0, readLength)
            post {
                if (mListener == null) {
                    return@post
                }
                mListener!!.onByte(mFile, mTotalByte, mDownloadByte)
                val progress = getProgressProgress(mTotalByte, mDownloadByte)
                // 只有下载进度发生改变的时候才回调此方法，避免引起不必要的 View 重绘
                if (progress != mDownloadProgress) {
                    mDownloadProgress = progress
                    mListener!!.onProgress(mFile, mDownloadProgress)
                    printLog(
                        mHttpRequest, mFile!!.path +
                                ", downloaded: " + mDownloadByte + " / " + mTotalByte +
                                ", progress: " + progress + " %"
                    )
                }
            }
        }
        closeStream(inputStream)
        closeStream(outputStream)
        val md5 = getFileMd5(openFileInputStream(mFile!!))
        if (!isEmpty(mMd5) && !mMd5.equals(md5, ignoreCase = true)) {
            // 文件 MD5 值校验失败
            throw MD5Exception("MD5 verify failure", md5!!)
        }
        post {
            if (mListener == null) {
                return@post
            }
            mListener!!.onComplete(mFile, false)
            mListener!!.onEnd(mFile)
        }
    }

    override fun onFailure(e: Exception) {
        printThrowable(mHttpRequest, e)
        // 打印错误堆栈
        val finalException = mHttpRequest.requestHandler!!.requestFail(mHttpRequest, e)
        if (finalException !== e) {
            printThrowable(mHttpRequest, finalException)
        }
        post {
            if (mListener == null) {
                return@post
            }
            mListener!!.onError(mFile, finalException)
            mListener!!.onEnd(mFile)
            printLog(mHttpRequest, mFile!!.path + " download completed")
        }
    }

    companion object {
        /** 文件 MD5 正则表达式  */
        private const val FILE_MD5_REGEX = "^[\\w]{32}$"
    }
}